File: function/test-function/test-function/sender.go
package main

import (
	"C"
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"log"
	"math/rand"
	"net"
	"net/http"
	"time"
)

const (
	dispatchURL = "http://faas-dispatch.openfaas.svc.cluster.local:8080/sendImage"
)

func getResults(ch chan<- image) {
	server := &http.Server{Addr: ":8089"}
	http.HandleFunc("/sendResult", func(w http.ResponseWriter, r *http.Request) {
		if r.Method != "POST" {
			http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
			return
		}

		var result image
		err := json.NewDecoder(r.Body).Decode(&result)
		if err != nil {
			http.Error(w, "Bad request", http.StatusBadRequest)
			return
		}
		ch <- result
		w.WriteHeader(http.StatusOK)
		go func() {
			err := server.Shutdown(context.Background())
			if err != nil {
				log.Fatal(err)
			}
		}()
	})
	server.ListenAndServe()
}

//export Test
func Test(i C.int, f C.int) {
	start := time.Now()
	ch := make(chan image)
	go sendImage(0, 1)
	go getResults(ch)
	result := <-ch
	fmt.Println("test", result)
	end := time.Since(start)
	fmt.Println(end)
}

func main() {
	rand.Seed(time.Now().UnixNano())
	// bs := 5

	Test(0, 6)

	return
}

func getContainerIP() string {
	addrs, err := net.InterfaceAddrs()
	if err != nil {
		fmt.Println(err)
		return ""
	}
	for _, addr := range addrs {
		ipnet, ok := addr.(*net.IPNet)
		if ok && !ipnet.IP.IsLoopback() && ipnet.IP.To4() != nil {
			return ipnet.IP.String()
		}
	}
	return ""
}

type image struct {
	Name string `json:"name"`
	From string `json:"from"`
	Data string `json:"data"`
}

func sendImage(i int, f int) {
	tmp := fmt.Sprintf("image%d.png", i)
	ip := getContainerIP()
	log.Printf("ip:%s", ip)
	imageData := image{tmp, ip, tmp}
	jsonData, err := json.Marshal(imageData)
	if err != nil {
	}
	resp, err := http.Post(dispatchURL, "application/json", bytes.NewBuffer(jsonData))
	if err != nil {
		fmt.Printf("Error sending image: %v\n", err)
		return
	}
	defer resp.Body.Close()
	fmt.Printf("Image sent with status: %s\n", resp.Status)
}

File: function/test-function/test-function/sender.py
import json
import random
import socket
import time
from http.server import BaseHTTPRequestHandler, HTTPServer
from threading import Thread
from urllib.request import Request, urlopen

dispatch_url = "http://10.244.0.18:5000/sendImage"


class ResultHandler(BaseHTTPRequestHandler):
    def do_POST(self):
        if self.path != "/sendResult":
            self.send_response(405)
            self.send_header("Content-type", "text/plain")
            self.end_headers()
            self.wfile.write(b"Method not allowed")
            return

        content_length = int(self.headers.get("Content-Length", 0))
        data = self.rfile.read(content_length)
        try:
            result = json.loads(data)
        except json.JSONDecodeError:
            self.send_response(400)
            self.send_header("Content-type", "text/plain")
            self.end_headers()
            self.wfile.write(b"Bad request")
            return

        self.send_response(200)
        self.end_headers()

        self.server.results.append(result)
        self.server.shutdown()


def get_results():
    server = HTTPServer(("", 8084), ResultHandler)
    server.results = []
    server.serve_forever()
    return server.results


def test(i, f):
    send_image(0, 1)
    results_thread = Thread(target=get_results, daemon=True)
    results_thread.start()
    results_thread.join()
    result = results_thread.result
    print("test", result)


def main():
    random.seed(time.time())
    bs = 5
    test(0, bs)


def get_container_ip():
    hostname = socket.gethostname()
    ip = socket.gethostbyname(hostname)
    return ip


class Image:
    def __init__(self, name, from_, data):
        self.name = name
        self.from_ = from_
        self.data = data

    def to_json(self):
        return json.dumps(self.__dict__)


def send_image(i, f):
    tmp = f"image{i}.png"
    image_data = Image(tmp, get_container_ip(), tmp)
    json_data = image_data.to_json()
    req = Request(dispatch_url, json_data.encode(), {"Content-Type": "application/json"})
    try:
        with urlopen(req) as resp:
            print(f"Image sent with status: {resp.status}")
    except Exception as e:
        print(f"Error sending image: {e}")
